src/wasm/wasm-binary.cpp

```diff
@@ -219,20 +219,35 @@ void WasmBinaryWriter::writeTypes() {
   if (indexedTypes.types.size() == 0) {
     return;
   }
-  // Count the number of recursion groups, which is always the number of
-  // elements in the type section. In non-isorecursive type systems, it is also
-  // equivalent to the number of types.
-  std::optional<RecGroup> lastGroup;
+  // Count the number of recursion groups, which is the number of elements in
+  // the type section. With nominal typing there is always one group and with
+  // equirecursive typing there is one group per type.
   size_t numGroups = 0;
-  for (auto type : indexedTypes.types) {
-    auto currGroup = type.getRecGroup();
-    numGroups += lastGroup != currGroup;
-    lastGroup = currGroup;
+  switch (getTypeSystem()) {
+    case TypeSystem::Equirecursive:
+      numGroups = indexedTypes.types.size();
+      break;
+    case TypeSystem::Nominal:
+      numGroups = 1;
+      break;
+    case TypeSystem::Isorecursive: {
+      std::optional<RecGroup> lastGroup;
+      for (auto type : indexedTypes.types) {
+        auto currGroup = type.getRecGroup();
+        numGroups += lastGroup != currGroup;
+        lastGroup = currGroup;
+      }
+    }
   }
   BYN_TRACE("== writeTypes\n");
   auto start = startSection(BinaryConsts::Section::Type);
   o << U32LEB(numGroups);
-  lastGroup = std::nullopt;
+  if (getTypeSystem() == TypeSystem::Nominal) {
+    // The nominal recursion group contains every type.
+    o << S32LEB(BinaryConsts::EncodedType::Rec)
+      << U32LEB(indexedTypes.types.size());
```

Comment:
Does this duplicate what happens on (new) line 259, in the case of nominal typing? Naively I'd guess 259 is also reached since lastGroup begins as null (different from currGroup) and the size of currGroup is greater than 1 (since in nominal typing, the group contains all the types (as the comment on 258 says). So at first reading it looks like we emit duplicate data here. If that's not the case, then I'd guess while the nominal group logically contains every type, we don't actually contain that data in currGroup (I assume due to https://github.com/WebAssembly/binaryen/blob/e7c4f9da6d4a98f58b6320b5769f653a6bf26d3e/src/wasm/wasm-type.cpp#L1449-L1451)? If so, a comment would be good, as this is all fairly complex with all the three type systems...

Reply:
Right, we internally represent nominal types as each being in their own group of size one since that is equivalent to not having a recursion group for most purposes. That means that new line 159 is never reached for nominal types. The comment on new lines 254-256 tries to get at that, but I'll make it more explicit.